File: TeeOkHttpClient.java
package com.example.tls.okhttp;

import com.example.tls.ssl.TeeSSLContextManager;
import com.example.tls.tee.TeeSignatureInterface;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.X509TrustManager;
import java.io.IOException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.concurrent.TimeUnit;

/**
 * 集成TEE的OkHttp客户端
 * 使用TEE进行TLS客户端认证
 */
public class TeeOkHttpClient {
    
    private final TeeSSLContextManager sslContextManager;
    private OkHttpClient okHttpClient;
    private final boolean enableLogging;
    
    public TeeOkHttpClient(TeeSignatureInterface teeInterface, boolean enableLogging) {
        this.sslContextManager = new TeeSSLContextManager(teeInterface);
        this.enableLogging = enableLogging;
    }
    
    /**
     * 构建OkHttp客户端
     * 
     * @param trustAllCerts 是否信任所有证书（仅用于测试）
     * @return 配置好的OkHttpClient
     * @throws Exception 构建过程中的异常
     */
    public OkHttpClient buildClient(boolean trustAllCerts) throws Exception {
        if (okHttpClient == null) {
            // 检查TEE可用性
            if (!sslContextManager.isTeeAvailable()) {
                throw new IllegalStateException("TEE is not available");
            }
            
            // 创建SSL上下文
            SSLContext sslContext = sslContextManager.createSSLContext(trustAllCerts);
            SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
            
            // 构建OkHttpClient
            OkHttpClient.Builder builder = new OkHttpClient.Builder()
                .sslSocketFactory(sslSocketFactory, getTrustManager(trustAllCerts))
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS);
            
            // 添加日志拦截器（如果启用）
            if (enableLogging) {
                HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
                loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.HEADERS);
                builder.addInterceptor(loggingInterceptor);
            }
            
            // 如果信任所有证书，禁用主机名验证（仅用于测试）
            if (trustAllCerts) {
                builder.hostnameVerifier((hostname, session) -> true);
            }
            
            okHttpClient = builder.build();
        }
        
        return okHttpClient;
    }
    
    /**
     * 执行GET请求
     * 
     * @param url 请求URL
     * @return 响应
     * @throws IOException 请求异常
     */
    public Response get(String url) throws IOException {
        if (okHttpClient == null) {
            throw new IllegalStateException("Client not built. Call buildClient() first.");
        }
        
        Request request = new Request.Builder()
            .url(url)
            .get()
            .build();
        
        return okHttpClient.newCall(request).execute();
    }
    
    /**
     * 执行POST请求
     * 
     * @param url 请求URL
     * @param body 请求体
     * @return 响应
     * @throws IOException 请求异常
     */
    public Response post(String url, okhttp3.RequestBody body) throws IOException {
        if (okHttpClient == null) {
            throw new IllegalStateException("Client not built. Call buildClient() first.");
        }
        
        Request request = new Request.Builder()
            .url(url)
            .post(body)
            .build();
        
        return okHttpClient.newCall(request).execute();
    }
    
    /**
     * 获取原始OkHttpClient实例
     * 
     * @return OkHttpClient实例
     */
    public OkHttpClient getClient() {
        return okHttpClient;
    }
    
    /**
     * 获取支持的签名算法
     * 
     * @return 算法数组
     */
    public String[] getSupportedAlgorithms() {
        return sslContextManager.getSupportedAlgorithms();
    }
    
    /**
     * 检查TEE是否可用
     * 
     * @return true if available
     */
    public boolean isTeeAvailable() {
        return sslContextManager.isTeeAvailable();
    }
    
    /**
     * 清理资源
     */
    public void cleanup() {
        sslContextManager.cleanup();
        
        if (okHttpClient != null) {
            // 关闭连接池
            okHttpClient.dispatcher().executorService().shutdown();
            okHttpClient.connectionPool().evictAll();
        }
    }
    
    /**
     * 获取信任管理器
     * 
     * @param trustAllCerts 是否信任所有证书
     * @return X509TrustManager
     */
    private X509TrustManager getTrustManager(boolean trustAllCerts) {
        if (trustAllCerts) {
            return new X509TrustManager() {
                @Override
                public void checkClientTrusted(X509Certificate[] chain, String authType) {
                    // 不做检查
                }
                
                @Override
                public void checkServerTrusted(X509Certificate[] chain, String authType) {
                    // 不做检查
                }
                
                @Override
                public X509Certificate[] getAcceptedIssuers() {
                    return new X509Certificate[0];
                }
            };
        } else {
            // 返回默认的信任管理器
            try {
                javax.net.ssl.TrustManagerFactory factory = 
                    javax.net.ssl.TrustManagerFactory.getInstance(
                        javax.net.ssl.TrustManagerFactory.getDefaultAlgorithm());
                factory.init((java.security.KeyStore) null);
                javax.net.ssl.TrustManager[] trustManagers = factory.getTrustManagers();
                
                for (javax.net.ssl.TrustManager tm : trustManagers) {
                    if (tm instanceof X509TrustManager) {
                        return (X509TrustManager) tm;
                    }
                }
            } catch (Exception e) {
                throw new RuntimeException("Failed to create default trust manager", e);
            }
            
            throw new RuntimeException("No X509TrustManager found");
        }
    }
}

File: android-tee-tls-example/src/main/java/com/example/teetls/TEEOkHttpClient.kt
package com.example.teetls

import android.util.Log
import okhttp3.*
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit

/**
 * 集成TEE的OkHttp客户端
 * 演示如何在OkHttp中使用基于TEE的自定义TLS引擎
 */
class TEEOkHttpClient private constructor() {
    
    companion object {
        private const val TAG = "TEEOkHttpClient"
        
        /**
         * 创建集成TEE的OkHttp客户端
         */
        fun create(): OkHttpClient {
            return try {
                // 初始化TEE密钥管理器
                val teeKeyManager = TEEKeyManager()
                
                // 创建自定义TLS Socket Factory
                val customSocketFactory = CustomTLSSocketFactory.create(teeKeyManager)
                
                // 创建日志拦截器
                val loggingInterceptor = HttpLoggingInterceptor { message ->
                    Log.d(TAG, "OkHttp: $message")
                }.apply {
                    level = HttpLoggingInterceptor.Level.HEADERS
                }
                
                // 创建自定义拦截器用于添加TEE签名头
                val teeSignatureInterceptor = TEESignatureInterceptor(teeKeyManager)
                
                // 构建OkHttp客户端
                OkHttpClient.Builder()
                    .sslSocketFactory(
                        customSocketFactory,
                        CustomTLSSocketFactory.createTrustManager()
                    )
                    .connectTimeout(30, TimeUnit.SECONDS)
                    .readTimeout(30, TimeUnit.SECONDS)
                    .writeTimeout(30, TimeUnit.SECONDS)
                    .addInterceptor(loggingInterceptor)
                    .addInterceptor(teeSignatureInterceptor)
                    .retryOnConnectionFailure(true)
                    .build()
                    .also {
                        Log.d(TAG, "TEE OkHttp客户端创建成功")
                    }
            } catch (e: Exception) {
                Log.e(TAG, "创建TEE OkHttp客户端失败", e)
                throw RuntimeException("创建TEE OkHttp客户端失败", e)
            }
        }
    }
}

/**
 * TEE签名拦截器
 * 为HTTP请求添加基于TEE的数字签名
 */
class TEESignatureInterceptor(
    private val teeKeyManager: TEEKeyManager
) : Interceptor {
    
    companion object {
        private const val TAG = "TEESignatureInterceptor"
        private const val SIGNATURE_HEADER = "X-TEE-Signature"
        private const val TIMESTAMP_HEADER = "X-TEE-Timestamp"
    }
    
    override fun intercept(chain: Interceptor.Chain): Response {
        val originalRequest = chain.request()
        
        try {
            // 生成时间戳
            val timestamp = System.currentTimeMillis().toString()
            
            // 构建签名数据
            val signatureData = buildSignatureData(originalRequest, timestamp)
            
            // 使用TEE进行签名
            val signature = teeKeyManager.signWithTEE(signatureData.toByteArray())
            val signatureBase64 = android.util.Base64.encodeToString(
                signature, 
                android.util.Base64.NO_WRAP
            )
            
            // 添加签名头到请求
            val signedRequest = originalRequest.newBuilder()
                .addHeader(SIGNATURE_HEADER, signatureBase64)
                .addHeader(TIMESTAMP_HEADER, timestamp)
                .build()
            
            Log.d(TAG, "已为请求添加TEE签名: ${originalRequest.url}")
            
            return chain.proceed(signedRequest)
        } catch (e: Exception) {
            Log.e(TAG, "TEE签名失败", e)
            // 如果签名失败，继续发送原始请求
            return chain.proceed(originalRequest)
        }
    }
    
    /**
     * 构建签名数据
     */
    private fun buildSignatureData(request: Request, timestamp: String): String {
        val method = request.method
        val url = request.url.toString()
        val headers = request.headers.toString()
        
        // 如果有请求体，包含请求体的哈希
        val bodyHash = request.body?.let { body ->
            val buffer = okio.Buffer()
            body.writeTo(buffer)
            val bodyBytes = buffer.readByteArray()
            android.util.Base64.encodeToString(
                java.security.MessageDigest.getInstance("SHA-256").digest(bodyBytes),
                android.util.Base64.NO_WRAP
            )
        } ?: ""
        
        return "$method|$url|$headers|$bodyHash|$timestamp"
    }
}